rust/core-lib/src/plugins.rs

```diff
@@ -0,0 +1,401 @@
+                .plugin_get_data(offset, max_size, rev)
+                .map(|data| Value::String(data))
+            }
+
+            Alert { msg } => {
+                self.buffers.lock().editor_for_view(view_id).unwrap()
+                .plugin_alert(&msg);
+                None
+            }
+        }
+    }
+}
+
+
+/// Starts a thread which collects editor updates and propagates them to plugins.
+///
+/// In addition to updates caused by user edits, updates can be caused by
+/// plugin edits. These updates arrive asynchronously. After being applied to
+/// the relevant buffer via an `Editor` instance, they need to be propagated
+/// back out to all interested plugins.
+///
+/// In order to avoid additional complexity in the model graph (e.g. giving each
+/// `Editor` a weak reference to the `PluginManager`) we instead give each
+/// `Editor` a tx end of an `mpsc::channel`. As plugin updates are generated,
+/// they are sent over this channel to a receiver running in another thread,
+/// which forwards them to interested plugins.
+pub fn start_update_thread<W: Write + Send + 'static>(
+    rx: mpsc::Receiver<(ViewIdentifier, PluginUpdate, usize)>,
+    plugins: Arc<Mutex<PluginManager<W>>>)
+{
+    thread::spawn(move ||{
+        loop {
+            let (view_id, update, undo_group) = rx.recv().unwrap();
+            plugins.lock().unwrap().update(&view_id, update, undo_group);
+        }
+    });
+    //FIXME: any reason not to drop the handle, here?
+}
+
+//TODO: Much of this might live somewhere else, and be shared with the plugin lib.
+
+// ============================================================================
+// Plugin RPC Types
+// ============================================================================
+
+/// An update event sent to a plugin.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PluginUpdate {
+    start: usize,
+    end: usize,
+    new_len: usize,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    text: Option<String>,
+    rev: usize,
+    edit_type: String,
+    author: String,
+}
+
+impl PluginUpdate {
+    pub fn new(start: usize, end: usize, new_len: usize, rev: usize,
+               text: Option<String>, edit_type: String, author: String) -> Self {
+        PluginUpdate {
+            start: start,
+            end: end,
+            new_len: new_len,
+            text: text,
+            rev: rev,
+            edit_type: edit_type,
+            author: author
+        }
+    }
+}
+
+/// An simple edit, received from a plugin.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PluginEdit {
+    pub start: u64,
+    pub end: u64,
+    pub rev: u64,
+    pub text: String,
+    /// the edit priority determines the resolution strategy when merging
+    /// concurrent edits. The highest priority edit will be applied last.
+    pub priority: u64,
+    /// whether the inserted text prefers to be to the right of the cursor.
+    pub after_cursor: bool,
```

Comment:
This will probably be subsumed by "priority", but I think that can wait until we actually implement that; it will require a bunch of changes.